lib/shared-method.js

```diff
@@ -303,30 +304,37 @@ function coerceAccepts(uarg, desc) {
     });
   }
 
+  var argIsNullOrUndefined = uarg === null || uarg === undefined;
   var actualType = SharedMethod.getType(uarg);
 
   // convert values to the correct type
   // TODO(bajtos) Move conversions to HttpContext (and friends)
   // SharedMethod should only check that argument values match argument types.
   var conversionNeeded = targetType !== 'any' &&
-    actualType !== 'undefined' &&
+    !argIsNullOrUndefined &&
     actualType !== targetType;
 
   if (conversionNeeded) {
-    // JSON.parse can throw, so catch this error.
-    try {
-      uarg = convertValueToTargetType(name, uarg, targetType);
-      actualType = SharedMethod.getType(uarg);
-    } catch (e) {
-      var message = util.format('invalid value for argument \'%s\' of type ' +
-        '\'%s\': %s. Received type was %s. Error: %s',
-        name, targetType, uarg, typeof uarg, e.message);
-      throw new badArgumentError(message);
+    if (Dynamic.canConvert(targetType)) {
+      // Convert input so it always matches the method definition.
+      // The user can also define new dynamic types.
+      uarg = new Dynamic(uarg).to(targetType);
+    } else {
+      // JSON.parse can throw, so catch this error.
+      try {
+        uarg = convertValueToTargetType(name, uarg, targetType);
+      } catch (e) {
+        var message = util.format('Invalid value for argument \'%s\' of type ' +
+          '\'%s\'. Received type was %s. Error: %s',
+          name, targetType, typeof uarg, e.message);
+        throw new badArgumentError(message);
```

Comment:
I am reluctant to remove request values from error messages. By including the request value, we make troubleshooting easier, e.g. when one doesn't have access to request bodies, only the error messages.

We already had a user asking to include values in validation errors, see https://github.com/strongloop/loopback-datasource-juggler/pull/389

> ```js
method.invoke('ctx', { obj: '<script>alert(1)</script>' }, function(err) {
```

The use case where the request value contains `<script>` element should be IMO handled by the client processing the responses, it's the client responsibility to html-encode any error messages before rendering them in the UI.

I am also not convinced that the argument value is the only place that can be abused for code injection, i.e. this change may be giving a false sense of security. However, if you insist, then I am willing to consider adding a feature flag to exclude request values from error messages, the flag should be disabled by default.

@raymondfeng @ritch thoughts?